controller/api/public/prometheus.go

```diff
@@ -135,10 +138,15 @@ func (s *grpcServer) getPrometheusMetrics(ctx context.Context, requestQueryTempl
 
 	// kick off asynchronous queries: request count queries + 3 latency queries
 	for pt, requestQueryTemplate := range requestQueryTemplates {
+		var query string
+		if pt == promTCPConnections {
+			query = fmt.Sprintf(requestQueryTemplate, labels, groupBy)
+		} else {
+			query = fmt.Sprintf(requestQueryTemplate, labels, timeWindow, groupBy)
+		}
+
 		go func(typ promType, template string) {
-			// success/failure counts
-			requestsQuery := fmt.Sprintf(template, labels, timeWindow, groupBy)
-			resultVector, err := s.queryProm(ctx, requestsQuery)
+			resultVector, err := s.queryProm(ctx, query)
```

Comment:
I'm pretty sure we want to continue to build the `query` variable inside this goroutine, or otherwise we risk running into a closure issue [like the one that's described here](https://blog.cloudflare.com/a-go-gotcha-when-closures-and-goroutines-collide/). Can we do something like this:

```diff
diff --git a/controller/api/public/prometheus.go b/controller/api/public/prometheus.go
index 884e49ca..a1143307 100644
--- a/controller/api/public/prometheus.go
+++ b/controller/api/public/prometheus.go
@@ -138,14 +138,13 @@ func (s *grpcServer) getPrometheusMetrics(ctx context.Context, requestQueryTempl
 
 	// kick off asynchronous queries: request count queries + 3 latency queries
 	for pt, requestQueryTemplate := range requestQueryTemplates {
-		var query string
-		if pt == promTCPConnections {
-			query = fmt.Sprintf(requestQueryTemplate, labels, groupBy)
-		} else {
-			query = fmt.Sprintf(requestQueryTemplate, labels, timeWindow, groupBy)
-		}
-
 		go func(typ promType, template string) {
+			var query string
+			if typ == promTCPConnections {
+				query = fmt.Sprintf(template, labels, groupBy)
+			} else {
+				query = fmt.Sprintf(template, labels, timeWindow, groupBy)
+			}
 			resultVector, err := s.queryProm(ctx, query)
 
 			resultChan <- promResult{
```

Reply:
great catch! as talked about in person, I've passed in the query string to the goroutine instead!